### PR TITLE
ceph: Allow removal of arbitrary osds on pvcs and simplify pvc names

### LINKED ISSUE #6287: Prevent rook from recreating automatically removed OSDs from PVCs
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
Prevent rook from recreating automatically removed OSDs from PVCs

**What is use case behind this feature:**
In cloud environments it would be nice to be able to remove any OSD from a PVC without having rook recreate the removed
OSD automatically minutes later. This doesnt happen if the removed OSD is the one related to the highest index PVC affected.
IMHO that should be the case for any removed OSD from a PVC, like it would behave in bare metal environments.
**Environment**:
rook 1.4.4
Kubernetes 1.18.8
DigitalOcean managed Kubernetes

### PR DESCRIPTION
**Pull request by travisn**
**Description of your changes:**
OSDs on PVCs previously were always replaced with PVCs of a given name. For on-prem scenarios where the OSDs might need to be removed instead of replaced, the operator now allows holes to exist in the PVC index names as long as there are a sufficient number of PVCs to meet the criteria for the deviceSet.count.

At the same time, we use predictable PVC names for the ODSs such as rook-ceph-osd-set1-data-3 as there is no need for a generated pvc suffix name

**Which issue is resolved by this Pull Request:**
Resolves #6287 #6948 

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: Documentation/ceph-osd-mgmt.md [hunk 1]
```diff
@@ -67,11 +67,11 @@ If you are using `useAllDevices: true`, no change to the CR is necessary.
...
 
-If you want to reduce the storage in your cluster, a healthy PVC-based OSD can be removed as described in this section.
+To reduce the storage in your cluster or remove a failed OSD on a PVC:
 
 1. Shrink the number of OSDs in the `storageClassDeviceSet` in the CephCluster CR.
    - `kubectl -n rook-ceph patch CephCluster rook-ceph --type=json -p '[{"op": "replace", "path": "/spec/storage/storageClassDeviceSets/0/count", "value":<desired number>}]'`
-   - Reduce the `count` of the OSDs to the desired number. Rook will not take any action to automatically remove the extra OSD(s), but will effectively stop managing the orphaned OSD.
+   - Reduce the `count` of the OSDs to the desired number. Rook will not take any action to automatically remove the extra OSD(s).
 2. Identify the orphaned PVC that belongs to the orphaned OSD.
    - The orphaned PVC will have the highest index among the PVCs for the device set.
```
Comment (satoru-takeuchi) [inline Documentation/ceph-osd-mgmt.md:76]: This line is not needed anymore since we can remove any OSD having an arbitrary index.

Comment (satoru-takeuchi) [reply Documentation/ceph-osd-mgmt.md:76]: And it's necessary to change the following line since PVC names don't have suffix anymore.

> For example if the device set is named set1 and the count was reduced from 3 to 2, the orphaned PVC would have the index 2 and might be named set1-2-data-vbwcf

**Comment by satoru-takeuchi** [comment]
I'm under review. Verifying many patterns as follows in two OSD environment.

a. Remove OSD 0 and add it again # making holes
b. remove OSD 1 and add it again # doesn't make holes

File: Documentation/ceph-osd-mgmt.md [hunk 2]
```diff
@@ -81,6 +81,9 @@ If you want to reduce the storage in your cluster, a healthy PVC-based OSD can b
    - `kubectl -n rook-ceph get pod -l ceph.rook.io/pvc=<orphaned-pvc> -o yaml | grep ceph-osd-id`
    - For example, this might return: `ceph-osd-id: "0"`
 
+If you later increase the count in the device set, note that the operator will create PVCs with a new index rather than using the
+same PVC names that were removed in these steps.
```
Comment (satoru-takeuchi) [inline Documentation/ceph-osd-mgmt.md:85]: In two OSDs (OSD0 and OSD1) environment, I confirmed that OSD1's PVC name became `rook-ceph-osd-set1-data-1` after removing OSD1 and adding it again. It's probably because the index is the highest existing number + 1. So it's not always incremented.

File: pkg/operator/ceph/cluster/osd/deviceSet.go [hunk 3]
```diff
@@ -27,100 +28,135 @@ import (
...
+			pvcTemplate.Name = bluestorePVCData
+		}
+
+		pvc, err := c.createDeviceSetPVC(existingPVCs, deviceSet.Name, pvcTemplate, setIndex)
+		if err != nil {
+			config.addError("failed to create osd for device set %q for count %d. %v", deviceSet.Name, setIndex, err)
+			continue
+		}
+
+		// The PVC type must be from a predefined set such as "data" and "metadata". These names must be enforced if the wal/db are specified
```
Comment (satoru-takeuchi) [inline pkg/operator/ceph/cluster/osd/deviceSet.go:112]: We can also set "wal".

File: pkg/operator/ceph/cluster/osd/deviceset_test.go [hunk 4]
```diff
@@ -83,10 +80,105 @@ func testPrepareDeviceSets(t *testing.T, setTemplateName bool) {
...
+	pvc, err := clientset.CoreV1().PersistentVolumeClaims(cluster.clusterInfo.Namespace).Get(ctx, "rook-ceph-osd-mydata-data-2", metav1.GetOptions{})
+	assert.NoError(t, err)
+	assert.NotNil(t, pvc)
+	pvc, err = clientset.CoreV1().PersistentVolumeClaims(cluster.clusterInfo.Namespace).Get(ctx, "rook-ceph-osd-mydata-metadata-2", metav1.GetOptions{})
+	assert.NoError(t, err)
+	assert.NotNil(t, pvc)
+	pvc, err = clientset.CoreV1().PersistentVolumeClaims(cluster.clusterInfo.Namespace).Get(ctx, "rook-ceph-osd-mydata-wal-2", metav1.GetOptions{})
+	assert.NoError(t, err)
+	assert.NotNil(t, pvc)
+}
```
Comment (satoru-takeuchi) [inline pkg/operator/ceph/cluster/osd/deviceset_test.go:171]: Making two tests are better. One is for making a hole case that removes OSD0 (it already exists). The other is for the non-hole case (it doesn't exist yet).

**Comment by satoru-takeuchi** [comment]
@travisn Review is done!

File: Documentation/ceph-osd-mgmt.md (hunk 2 above)
Comment (travisn) [reply Documentation/ceph-osd-mgmt.md:85]: Yes, I'll clarify. If the highest index is deleted, it will be re-used.

File: pkg/operator/ceph/cluster/osd/deviceset_test.go (hunk 4 above)
Comment (travisn) [reply pkg/operator/ceph/cluster/osd/deviceset_test.go:171]: I think those test cases are already covered, right? Are you just suggesting to split the new test into two functions? I'd prefer to keep it as one unless it's a different test I missed.

File: Documentation/ceph-osd-mgmt.md [hunk 5]
```diff
@@ -67,20 +67,21 @@ If you are using `useAllDevices: true`, no change to the CR is necessary.
 
 ### PVC-based cluster
 
-If you want to reduce the storage in your cluster, a healthy PVC-based OSD can be removed as described in this section.
+To reduce the storage in your cluster or remove a failed OSD on a PVC:
 
 1. Shrink the number of OSDs in the `storageClassDeviceSet` in the CephCluster CR.
    - `kubectl -n rook-ceph patch CephCluster rook-ceph --type=json -p '[{"op": "replace", "path": "/spec/storage/storageClassDeviceSets/0/count", "value":<desired number>}]'`
```
Comment (BlaineEXE) [inline Documentation/ceph-osd-mgmt.md:73]: It might be good to note that the `0` here is the index of the particular StorageClassDeviceSet if they have multiple.

File: pkg/operator/ceph/cluster/osd/deviceset_test.go (hunk 4 above)
Comment (BlaineEXE) [reply pkg/operator/ceph/cluster/osd/deviceset_test.go:171]: I think the suggestion is to add a second test that ensures that if the highest-indexed ID is deleted, it gets reused when scaled back up. This might be a new test like `TestPrepareDeviceSetWithoutHolesInPVC`

Comment (travisn) [reply pkg/operator/ceph/cluster/osd/deviceset_test.go:171]: I think that this is implicit already in the tests where it verified that the first available index (2) was used to add a new osd. If there are no holes in the index, it's just the base case compared to if there were holes.

Comment (BlaineEXE) [reply pkg/operator/ceph/cluster/osd/deviceset_test.go:171]: I do agree that it's not necessary, but I think it's still valuable to add that expectation to the test. In some respect, it is testing the corner case of _remove and re-create the **last** list item_ just as the first test does the same for the _**first** item_ where first and last items in a list are often corner cases. If I were to write the tests, I would probably test behavior with zero OSDs present, with one OSD present, and all 3 cases of removing and re-creating one item from a 3-item list to be sure there weren't unexpected logic issues.

Also, I have been advised before that the surest way to determine what code does is to examine the tests.

File: pkg/operator/ceph/cluster/osd/deviceSet.go [hunk 6]
```diff
@@ -162,50 +198,59 @@ func (c *Cluster) updatePVCIfChanged(desiredPVC *v1.PersistentVolumeClaim, curre
...
 		pvcLabels[k] = v
 	}
 
+	// pvc naming format rook-ceph-osd-<deviceSetName>-<SetNumber>-<PVCIndex>
 	return &v1.PersistentVolumeClaim{
 		ObjectMeta: metav1.ObjectMeta{
-			GenerateName: pvcGenerateName,
-			Labels:       pvcLabels,
-			Annotations:  pvcTemplate.Annotations,
+			Name:        fmt.Sprintf("%s-%s", AppName, pvcID),
```
[the remark this answers is not in the document]
Comment (travisn) [reply pkg/operator/ceph/cluster/osd/deviceSet.go:212]: Ok that's a good catch! To be safe to avoid re-using the same folder I will revert to generate the name again. It was just a nice-to-have for predictable names.

File: pkg/operator/ceph/cluster/osd/deviceset_test.go (hunk 4 above)
Comment (travisn) [reply pkg/operator/ceph/cluster/osd/deviceset_test.go:171]: I'll add one more step to the tests that will start with count of 1, then increase to 2, rather than just starting with 2. I believe this will cover the test hole you're looking for.

**Review by satoru-takeuchi — APPROVED** [review]
